### PR TITLE
Use the new Vertx local context storage

### PR DESCRIPTION
**Pull request by franz1981**
This is to keep up with the Vertx API changes for https://github.com/eclipse-vertx/vert.x/commit/88cf77bfe4f44c3dc5ee89d65b8891674088de46, although I'm not very happy with the current state (that's why is in draft):
1. `VertxLocalsHelper.` is "stealthy" using the new API by intercepting when users are passing `ContextLocal`-like keys, which means performing 2 instanceof checks (one being an interface check too, which is not yet safe with JDK till 21...)
2. As written in the `TODO` parts, we have some ordering issue: we need the keys statically defined as `ACCESS_TOGGLE_KEY` to be initialized before instantiating the `Vertx` instance has to use them, due to https://github.com/eclipse-vertx/vert.x/blob/4.5.7/src/main/java/io/vertx/core/impl/VertxImpl.java#L194 which is forcing every new context creation to dedicate the fixed amount of "optimized" keys in the context's local storage
3. The new `AccessMode` I've defined is good enough that having a new method in `AccessMode::putRelease`, would have been better IMO (@vietj )

For @mkouba and @Ladicek : the new "optimized" API for local storage is double edge sword because it needs some care on init ordering, need some care in dev mode (I think, @geoand ? @sanne ?) and requires to create local "permanent" keys only for the ones we're 100% are going to be stored on each context (and I need some help from @Ladicek and @mkouba for such, I have no IDEA!)

Additionally there are some weird things in the original API (but is ok, the purpose of dog fooding is indeed to try things and found how to use them!): @vietj see https://github.com/quarkusio/quarkus/compare/main...franz1981:quarkus:local_ctx?expand=1#diff-af571e338cf42e877b9b99b70840f0842b293365f18c24d5fb24c930e5b0b064R64: this one is just wrong, because we cannot query all the local storage `ContextLocal`'s values stored in the source context...we cannot do it unless we have a new method as `ContextInternal::addLocals(ContextInternal from, AccessMode accessMode)`

### CODE REVIEW

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusAccessModes.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    public static final AccessMode ACQUIRE_RELEASE = new AccessMode() {
+        @Override
+        public Object get(AtomicReferenceArray<Object> locals, int idx) {
+            return locals.get(idx);
+        }
+
+        @Override
+        public void put(AtomicReferenceArray<Object> locals, int idx, Object value) {
+            locals.lazySet(idx, value);
```
Comment (franz1981) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/QuarkusAccessModes.java:18]: lazySet is a `putRelease` in the new JMM, which means that hasn't total ordering effects

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java [hunk 2]
```diff
@@ -41,20 +56,47 @@ public <T extends InjectableContext.ContextState> CurrentContext<T> create(Class
         return new VertxCurrentContext<>(key);
     }
 
-    /**
-     *
-     * @return an unmodifiable list of used keys
-     */
-    public List<String> keys() {
-        return unmodifiableKeys;
+    public ContextInternal duplicateContext(ContextInternal vertxContext) {
```
Comment (mkouba) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:59]: The name of the method is misleading as we don't always duplicate the context...

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java [hunk 3]
```diff
@@ -14,23 +18,34 @@
...
         this.keys = new CopyOnWriteArrayList<>();
-        // We do not want to allocate a new object for each VertxCurrentContextFactory#keys() invocation
-        this.unmodifiableKeys = Collections.unmodifiableList(keys);
+        this.requestScopedKeyCreated = new AtomicBoolean();
     }
 
     @Override
     public <T extends InjectableContext.ContextState> CurrentContext<T> create(Class<? extends Annotation> scope) {
+        if (scope == RequestScoped.class) {
+            if (!requestScopedKeyCreated.compareAndExchange(false, true)) {
```
Comment (mkouba) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:42]: Heh, so we have a special case for `@RequestScoped` again? 

What's the diff betweeen `compareAndSet()` and `compareAndExchange()` anyway?

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java [hunk 4]
```diff
@@ -41,20 +56,47 @@ public <T extends InjectableContext.ContextState> CurrentContext<T> create(Class
...
-     *
-     * @return an unmodifiable list of used keys
-     */
-    public List<String> keys() {
-        return unmodifiableKeys;
+    public ContextInternal duplicateContext(ContextInternal vertxContext) {
+        List<String> keys = this.keys;
+        // TODO optimize it to do it right!
+        ConcurrentMap<Object, Object> local = vertxContext.localContextData();
+        if (containsScopeKey(keys, local)) {
```
Comment (mkouba) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:63]: This would return `false` if `local` only contains the `REQUEST_SCOPED_LOCAL_KEY`, or?

The list of keys does not contain the special key for `@RequestScoped`.

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java [hunk 5]
```diff
@@ -14,23 +18,34 @@
...
 import io.vertx.core.Context;
 import io.vertx.core.Vertx;
+import io.vertx.core.impl.ContextInternal;
+import io.vertx.core.impl.ContextLocalImpl;
 
 public class VertxCurrentContextFactory implements CurrentContextFactory {
 
     private static final String LOCAL_KEY_PREFIX = "io.quarkus.vertx.cdi-current-context";
+    // TODO VertxImpl should be allocated AFTER getting here, to make it work!
+    private static final ContextLocalImpl<? extends ContextState> REQUEST_SCOPED_LOCAL_KEY = new ContextLocalImpl<>();
```
Comment (mkouba) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:28]: So the `ContextLocalImpl` can be used as a key for local data? Shouldn't we use `io.vertx.core.spi.context.storage.ContextLocal.get(Context)` instead? I have to admin that I find the API quite confusing...

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java (hunk 3 above)
Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:42]: Ops, good point, I didn't meant to use such but compareAndSet: I will fix it!
Related being a special case, it really is, unless I can blindly trust that who is calling this method deserve to have their local keys to always require a slot in every context created by vertx (which is the whole point of the new local storage feature)

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java (hunk 4 above)
Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:63]: I have used the atomic boolean remove from the other context the request scoped key, if any

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java (hunk 2 above)
Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:59]: Yep, I wasn't happy about it, so I can reword it to make it clear

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java (hunk 5 above)
Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:28]: Depends if you wanna have optimized access or not, I suppose. And yes, indeed using this API within quarkus hopefully improve both (that's why I pinged @vietj here)

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java (hunk 4 above)
Comment (mkouba) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:63]: My point is that `containsScopeKey(keys, local)` only checks the list of keys. So if no string key is used (`REQUEST_SCOPED_LOCAL_KEY` is not added to the list) then this method will never create a new duplicated context...

**Comment by franz1981** [comment]
PTAL @mariofusco @mkouba 

Note: I'm still concerned about  the ordering by which these statically defined "keys" are generated, in order to happen before `Vertx` is created: any suggestion on how to do it the quarkus-way?

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java [hunk 6]
```diff
@@ -41,20 +55,49 @@ public <T extends InjectableContext.ContextState> CurrentContext<T> create(Class
...
+        }
+        return vertxContext;
+    }
+
+    private boolean containsAnyCreatedScopeKeys(ContextInternal vertxContext) {
+        boolean requestScopedKeyCreated = this.requestScopedKeyCreated.get();
+        if (requestScopedKeyCreated && vertxContext.getLocal(REQUEST_SCOPED_LOCAL_KEY) != null) {
+            return true;
+        }
+        if (keys.isEmpty()) {
```
Comment (mariofusco) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:77]: Not sure what is the most common scenario tbh, but maybe it could help performances to check this as first thing, bringing it at the very beginning of this method?

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java [hunk 7]
```diff
@@ -41,20 +55,49 @@ public <T extends InjectableContext.ContextState> CurrentContext<T> create(Class
...
-    /**
-     *
-     * @return an unmodifiable list of used keys
-     */
-    public List<String> keys() {
-        return unmodifiableKeys;
+    public ContextInternal duplicateContextIfContainsAnyCreatedScopeKeys(ContextInternal vertxContext) {
+        if (containsAnyCreatedScopeKeys(vertxContext)) {
+            // Duplicate the context, copy the data, remove the request context
+            vertxContext = vertxContext.duplicate();
```
Comment (mariofusco) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:61]: I believe this is a bug: you're reassigning the `vertxContext` variable, but if the intent of the subsequent line is also copying the `localContextData` from the old to the new context, this of course won't work.

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java (hunk 6 above)
Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:76]: Usually the most common thing is that `requestScopeKeyCreated` is true - it is basically always done AFAIK

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/storage/QuarkusLocalStorageKeyVertxServiceProvider.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import io.quarkus.vertx.core.runtime.context.VertxContextSafetyToggle;
+import io.vertx.core.impl.VertxBuilder;
+import io.vertx.core.spi.VertxServiceProvider;
+import io.vertx.core.spi.context.storage.ContextLocal;
+
+/**
+ * This provider exists with the sole purpose of reliably get the optimized local keys created before
+ * the Vertx instance is created.
+ */
+public class QuarkusLocalStorageKeyVertxServiceProvider implements VertxServiceProvider {
```
Comment (franz1981) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/storage/QuarkusLocalStorageKeyVertxServiceProvider.java:13]: @geoand @cescoffier I don't like to make spi loading to happen just for this TBH and I could, maybe, hack the `Vertx` recorder to just perform the init at the right time, but IDK what's the cost of this, usually. Any suggestion?

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java (hunk 7 above)
Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java]: Thanks @mariofusco , yep I've optimized without thinking here; this now should be fine-ish (check the new TODO on the code on why it is not!)

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java [hunk 9]
```diff
@@ -41,20 +54,52 @@ public <T extends InjectableContext.ContextState> CurrentContext<T> create(Class
...
-     */
-    public List<String> keys() {
-        return unmodifiableKeys;
+    public ContextInternal duplicateContextIfContainsAnyCreatedScopeKeys(ContextInternal vertxContext) {
+        if (!containsAnyCreatedScopeKeys(vertxContext)) {
+            return vertxContext;
+        }
+        // Duplicate the context, copy the data, remove the request context
+        var duplicateCtx = vertxContext.duplicate();
+        // TODO this is not copying any ContextLocal<?> from the original context to the new one!
```
Comment (franz1981) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:63]: @vietj @mariofusco This is just not right :"( : in short, given that there is no `ContextInternal`'s API to retrieve the existing `ContextLocal` keys (apart from the `ConcurrentHashMap` at index 0!!!!), 
IDK which ones should be copied and which not....

Having a `ContextInternal::putLocal(AccessMode mode, ContextInternal from)` method would be great here, because can skip any collection allocation e.g. `ContextInternal::getContextLocals` would force some allocation to happen.

The same (even better) could be achieved with `ContextInternal::duplicate(bool copyLocalStorage)` which can further avoid allocating any `ConcurrentHashMap` (happening at `ContextInternal::localContextData` call, regadless we have nothing in :"( )

wdyt @mariofusco do you want to send such PR to vertx? 
@vietj do you agree on this?

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxLocalsHelper.java [hunk 10]
```diff
@@ -22,6 +24,11 @@ public static void throwOnRootContextAccess() {
     public static <T> T getLocal(ContextInternal context, Object key) {
         if (VertxContext.isDuplicatedContext(context)) {
             // We are on a duplicated context, allow accessing the locals
+            if (key instanceof ContextLocalImpl<?>) {
```
Comment (franz1981) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxLocalsHelper.java:27]: @cescoffier I will probably add the new `ContextLocal`-based methods to avoid this terrible hack .-.
I made it in order to have the existing code in the toggle to just use the now-deprecated methods and still get the benefit of the new API, but given that these methods are going to disappear to favour the new API, I'll probably tackle this on the current PR already, wdyt?

File: extensions/vertx/runtime/src/main/resources/META-INF/services/io.vertx.core.spi.VertxServiceProvider [hunk 11]
```diff
@@ -0,0 +1 @@
+io.quarkus.vertx.runtime.storage.QuarkusLocalStorageKeyVertxServiceProvider
```
Comment (franz1981) [inline extensions/vertx/runtime/src/main/resources/META-INF/services/io.vertx.core.spi.VertxServiceProvider:1]: @mariofusco kudos to you, it was a proper SPI!

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java (hunk 5 above)
Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java]: @vietj wdyt? suggestions?

Comment (vietj) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java]: I do not recommend using ContextLocalImpl directly

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/context/VertxContextSafetyToggle.java [hunk 12]
```diff
@@ -50,6 +52,12 @@ public final class VertxContextSafetyToggle {
     private static final boolean UNRESTRICTED_BY_DEFAULT = Boolean.getBoolean(UNRESTRICTED_BY_DEFAULT_PROPERTY);
     private static final boolean FULLY_DISABLED = Boolean.getBoolean(FULLY_DISABLE_PROPERTY);
 
+    public static ContextLocal<Boolean> registerAccessToggleKey() {
+        if (FULLY_DISABLED)
+            return null;
+        return ContextLocal.registerLocal(Boolean.class);
```
Comment (franz1981) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/context/VertxContextSafetyToggle.java:58]: @viej is this ok?

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/storage/QuarkusLocalStorageKeyVertxServiceProvider.java (hunk 8 above)
Comment (cescoffier) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/storage/QuarkusLocalStorageKeyVertxServiceProvider.java:13]: The question is: can we preload this at build time (static init). It would be interesting to do that and just use the result at runtime. I don't believe there is a good reason to change the SPI at runtime.

Comment (geoand) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/storage/QuarkusLocalStorageKeyVertxServiceProvider.java:13]: I think we can skip the service loading in this case and just use this class,  no?

Comment (cescoffier) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/storage/QuarkusLocalStorageKeyVertxServiceProvider.java:13]: Yeah, I don't believe we should allow any other implementation.

Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/storage/QuarkusLocalStorageKeyVertxServiceProvider.java:13]: thanks both, yeah, I'm not very happy about it: so i can just have a not vertx service provider, but an holder class which is going to be initialized in the vertx recorder before vertx instances are allocated, wdyt?

Comment (geoand) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/storage/QuarkusLocalStorageKeyVertxServiceProvider.java:13]: Sounds fine to me

**Comment by franz1981** [comment]
This PR will remain in draft till https://github.com/eclipse-vertx/vert.x/pull/5215 is going to be into Vertx 4.x because ATM we have no way to duplicate "optimized" keys while creating a new duplicate context.

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/context/VertxContextSafetyToggle.java (hunk 12 above)
Comment (vietj) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/context/VertxContextSafetyToggle.java:58]: this is not how it should be used imho

**Comment by franz1981** [comment]
This PR is still blocked by the mentioned https://github.com/eclipse-vertx/vert.x/pull/5215 one, which should be ported to 4.x as well.

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/context/VertxContextSafetyToggle.java (hunk 12 above)
Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/context/VertxContextSafetyToggle.java:58]: what do you mean?
In our context (quarkus) we can disable specific keys, in a stable way (via the sys property), meaning that we don't have to register any local key there. 
Do you have another way to provide a similar behaviour?

**Comment by franz1981** [comment]
@cescoffier https://github.com/eclipse-vertx/vertx-grpc/blob/main/vertx-grpc-context-storage/src/main/java/io/grpc/override/ContextStorageOverride.java#L27 probably I need to send a PR just to fix this which is should be broken in master

**Comment by cescoffier** [comment]
@franz1981 should we go forward with this, or should it wait Vertx 5?

**Comment by franz1981** [comment]
Hi @cescoffier I would love to move this forward but there is a missing "feature" in vertx 4.x i.e.  moving local storage values from duplicated ctx with a filter - or just been able to query it. 
In an hour I will attach the relevant issue which explain this, because the same applies to vertx 5: if we want to use the new API we need it on vertx

**Comment by cescoffier** [comment]
@franz1981 do you have the relevant issues?
